### PR TITLE
[CBRD-24149] Change the comparison spec for char-type and varchar-type.

### PR DESCRIPTION
**Pull request by beyondykk9**
http://jira.cubrid.org/browse/CBRD-24149

Since version 11.0 in char string comparison spec is as below:
    char-type vs. char-type => char-type comparison
    char-type vs. varchar-type => varchar-type comparison
    * commonly char-type string's trailing space is trimmed first before comparing.

In the index search, however, the above comparison spec is not  applied. The char-type and varchar-type comparison is done as like char-type and char-type (ignores trailing space).

### CODE REVIEW

**Comment by ctshim** [comment]
Is it unnecessary to trim separately in the mr_cmpval_string() and mr_cmpval_char() functions, and is it solved by appropriately adjusting the ti parameter when calling QSTR_COMPARE()?

Even so, isn't the code below necessary in the mr_cmpval_char() function?
```
if (type1 != DB_TYPE_CHAR && type1 != DB_TYPE_NCHAR && type2 != DB_TYPE_CHAR && type2 != DB_TYPE_NCHAR)
   {
        ti = false;
    }
```

**Comment by shparkcubrid** [comment]
I commented a test case that caused the problem in jira.

File: src/object/object_primitive.c [hunk 1]
```diff
@@ -11134,33 +11134,10 @@ mr_cmpval_string (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int tot
 
   if (!ignore_trailing_space)
     {
-      int i;
-
-      if (type1 == DB_TYPE_CHAR || type1 == DB_TYPE_NCHAR)
+      if (type1 != DB_TYPE_CHAR && type1 != DB_TYPE_NCHAR && type2 != DB_TYPE_CHAR && type2 != DB_TYPE_NCHAR)
```
Comment (hgryoo) [inline src/object/object_primitive.c]:
```suggestion
      if (!TP_IS_FIXED_LEN_CHAR_TYPE (type1) && !TP_IS_FIXED_LEN_CHAR_TYPE (type2))
```

**Review by hyunikn — APPROVED** [review]

File: src/parser/csql_grammar.y [hunk 2]
```diff
@@ -27903,6 +27903,7 @@ pt_create_char_string_literal (PARSER_CONTEXT *parser, const PT_TYPE_ENUM char_t
         length = node->info.value.data_value.str->length;
 
         node->type_enum = char_type;
+/*
```
Comment (ctshim) [inline src/parser/csql_grammar.y]: In the query below, the tailing space of a varchar column is treated like a char column. Is your intention correct?
`select * from tbl where varchar_col = 'abc  '`

Comment (beyondykk9) [reply src/parser/csql_grammar.y]: yes, the policy of varchar-type constant is changed.

File: src/object/object_primitive.c (hunk 1 above)
Comment (beyondykk9) [reply src/object/object_primitive.c]: I agree with you.

File: src/parser/csql_grammar.y (hunk 2 above)
Comment (ctshim) [reply src/parser/csql_grammar.y]: It would be nice if you could write in a comment that you would like to comment on the policy change.

**Review by ctshim — APPROVED** [review]

**Review by hyunikn — COMMENTED** [review]
I heard that Cubrid regards a text literal as a varchar if it ends with trailing spaces and as a char if it does not, while Oracle always regards it as a char. Do we keep this policy unchanged?

File: src/parser/csql_grammar.y (hunk 2 above)
Comment (beyondykk9) [reply src/parser/csql_grammar.y]: Ok, I commented at JIRA as you said.

File: src/object/object_primitive.c [hunk 3]
```diff
@@ -11132,35 +11132,9 @@ mr_cmpval_string (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int tot
...
-	      if (string2[i - 1] != 0x20)
-		{
-		  break;
-		}
-	    }
-	  size2 = i;
-	}
-
-      ti = false;
+      ti = true;
```
Comment (hyunikn) [inline src/object/object_primitive.c]: We can now remove ti and use ignore_trailing_space instead.

Comment (beyondykk9) [reply src/object/object_primitive.c]: **ignore_trailing_space** is a static global and it is set only once, however, the variable **ti** is set whenever this routine is called. Therefore **ti** has a role to apply comparison method.

File: src/object/object_primitive.c [hunk 4]
```diff
@@ -12015,55 +11989,20 @@ mr_cmpval_char (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int total
...
-    }
-  else if (!ignore_trailing_space &&
-	   (type1 == DB_TYPE_STRING || type1 == DB_TYPE_VARNCHAR || type2 == DB_TYPE_STRING
-	    || type2 == DB_TYPE_VARNCHAR))
+  if (!ignore_trailing_space)
     {
-      int i;
-
-      if (type1 == DB_TYPE_CHAR || type1 == DB_TYPE_NCHAR)
+      if (do_coercion == 2)
```
Comment (hyunikn) [inline src/object/object_primitive.c]:
```suggestion
      if (do_coercion != 2 && 
          (!TP_IS_FIXED_LEN_CHAR_TYPE (type1) || !TP_IS_FIXED_LEN_CHAR_TYPE (type2)))
        {
          ti = false;
        }
```

File: src/object/object_primitive.c [hunk 5]
```diff
@@ -12015,55 +11989,20 @@ mr_cmpval_char (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int total
...
 	{
-	  for (i = size1; i > 1; i--)
+	  /* only for btree_get_prefix_sparator */
+	  if (TP_IS_FIXED_LEN_CHAR_TYPE (type1) || TP_IS_FIXED_LEN_CHAR_TYPE (type2))
 	    {
-	      if (string1[i - 1] != 0x20)
-		{
-		  break;
-		}
+	      ti = true;
```
Comment (hyunikn) [inline src/object/object_primitive.c]: ti is already true, and hence the above suggestion.

File: src/object/object_primitive.c [hunk 6]
```diff
@@ -12963,55 +12902,20 @@ mr_cmpval_nchar (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int tota
...
-   * we need to process enforcing no-ignore-trailing space.
-   */
-  else if (do_coercion == 3)
-    {
-      ti = false;
-    }
-  else if (!ignore_trailing_space &&
-	   (type1 == DB_TYPE_STRING || type1 == DB_TYPE_VARNCHAR || type2 == DB_TYPE_STRING
-	    || type2 == DB_TYPE_VARNCHAR))
+  if (!ignore_trailing_space)
```
Comment (hyunikn) [inline src/object/object_primitive.c:12898]: This if block can be erased because ti is already true.

File: src/object/object_primitive.c (hunk 3 above)
Comment (hyunikn) [reply src/object/object_primitive.c]: ti always has the same value as ignore_trailing_space at line 11140, and there is no other use of ti.

File: src/object/object_primitive.c (hunk 4 above)
Comment (beyondykk9) [reply src/object/object_primitive.c]: I understand your comment, but I keep the code and comment for readability.

File: src/object/object_primitive.c (hunk 3 above)
Comment (beyondykk9) [reply src/object/object_primitive.c]: Ok, I refactored the code as your comment.

**Pull request merged**